File: day4/second/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

func main() {
	input, err := os.Open("../input.txt")
	if err != nil {
		os.Exit(1)
	}
	defer input.Close()
	scanner := bufio.NewScanner(input)

	mat := [][]byte{}
	for scanner.Scan() {
		mat = append(mat, append([]byte{}, scanner.Bytes()...))
	}
	n, m := len(mat), len(mat[0])

	total := 0
	for r := 1; r < n-1; r++ {
		for c := 1; c < m-1; c++ {
			if mat[r][c] != 'A' {
				continue
			}
			str := string([]byte{mat[r-1][c-1], 'A', mat[r+1][c+1]})
			if str != "MAS" && str != "SAM" {
				continue
			}
			str = string([]byte{mat[r+1][c-1], 'A', mat[r-1][c+1]})
			if str != "MAS" && str != "SAM" {
				continue
			}
			total++
		}
	}

	fmt.Println(total)
}

File: day4/second/main.py
mat = []

with open("../input.txt") as f:
    for line in f:
        mat.append(line.strip())

n, m = len(mat), len(mat[0])

total = 0

for r in range(1, n-1):
    for c in range(1, m-1):
        if mat[r][c] != 'A':
            continue
        str = mat[r-1][c-1] + 'A' + mat[r+1][c+1]
        if str != 'MAS' and str != 'SAM':
            continue
        str = mat[r+1][c-1] + 'A' + mat[r-1][c+1]
        if str != 'MAS' and str != 'SAM':
            continue
        total += 1

print(total)